---
title: "Distribution Free Methods: Permutation Tests"
subtitle: "Quantitative Methods in Life Sciences"
author: 'Elizabeth King and Kevin Middleton'
output:
  ioslides_presentation:
    fig_width: 8
    css: styles.css
csl: evolution.csl
bibliography: Multivariate.bib
---

```{r setup, echo=FALSE, message=FALSE, warning=FALSE}
library(tidyverse)
library(cowplot)
library(readxl)
library(wesanderson)
knitr::opts_chunk$set(cache = TRUE)
```

## Optional background reading

Churchill, G. A., and R. W. Doerge. 1994. Empirical threshold values for quantitative trait mapping. Genetics 138:963-971.

## Genetic Mapping

- Over 11,000 linear models, one per position
- How many do we expect to be significant by chance?

```{r echo=FALSE}
load("../data/set_all.rda")
top <- 8
ggplot(set[set$Treatment == 'Life DR',], aes(x = Position, y = P)) +
  geom_rect(xmin = 66.3, ymin = -10, xmax = 120.3,
            ymax = top + 10, fill = 'aliceblue',
            color = 'aliceblue') +
  geom_rect(xmin = 174, ymin = -10, xmax = 221,
            ymax = top + 10, fill = 'aliceblue',
            color = 'aliceblue') +
  geom_line() +
  scale_x_continuous(
    breaks = c(0,66.3,120,174,221,277,33.15,98.3,145,205,249), 
                     labels = c(0, "66  0", 54, "108  0", 47, 103,
                                '\nX', '\n2L', '\n2R', '\n3L', '\n3R')) +
  theme(axis.ticks.x = element_blank()) +
  ylab(expression("-log"[10]*"(P-value)")) +
  xlab("Position (cM)") +
  geom_hline(yintercept = -log10(0.05), color = "coral")
```

## Genetic Mapping

What is the threshold where we expect to find at least one significant hit by chance alone 5% of the time?

- Find expectation of the random strength of association between genotype and phenotype by shuffling the phenotypes 
    - Iterate, perform 11k linear models for each
    - Choose maximum -log~10~P each time
    
```{r pl1, echo=FALSE, fig.height=2.5}
sub.d <- subset(set, Treatment == 'Life C')
dd <- data.frame('xx' = sub.d[which.max(sub.d$P), 'Position'], 
                 'yy' = sub.d[which.max(sub.d$P), 'P'])
top <- 8
ggplot(set[set$Treatment == 'Life C',], aes(x = Position, y = P)) +
  geom_rect(xmin = 66.3, ymin = -10, xmax = 120.3,
            ymax = top + 10, fill = 'aliceblue',
            color = 'aliceblue') +
  geom_rect(xmin = 174, ymin = -10, xmax = 221,
            ymax = top + 10, fill = 'aliceblue',
            color = 'aliceblue') +
  geom_line() +
  geom_point(data = dd, aes(xx, yy, color = 'coral'), size = 3) +
  scale_x_continuous(
    breaks = c(0, 66.3, 120, 174, 221, 277, 33.15, 98.3, 145, 205, 249), 
    labels = c(0, "66  0", 54, "108  0", 47, 103,
               '\nX', '\n2L', '\n2R', '\n3L', '\n3R')) +
  theme(axis.ticks.x = element_blank()) +
  ylab(expression("-log"[10]*"(P-value)")) +
  xlab("Position (cM)") +
  theme(legend.position = "none")
```

## Distribution of Maxima

```{r pl2, echo=FALSE}
load("../data/exPPP.rda")
ppp <- data.frame('MaxValue' = ppp)
ggplot(ppp, aes(x = MaxValue)) +
  geom_histogram(bins = 30) +
  geom_vline(xintercept = quantile(ppp$MaxValue, 0.95), color = 'coral')
```

## Genetic Mapping

```{r pl3, echo=FALSE}
ggplot(set[set$Treatment == 'Life DR', ], aes(x = Position, y = P)) +
  geom_rect(xmin = 66.3, ymin = -10, xmax = 120.3,
            ymax = top + 10, fill = 'aliceblue',
            color = 'aliceblue') +
  geom_rect(xmin = 174, ymin = -10, xmax = 221,
            ymax = top + 10, fill = 'aliceblue',
            color = 'aliceblue') +
  geom_line() +
  geom_hline(yintercept = quantile(ppp$MaxValue, 0.95), color = 'coral') +
  scale_x_continuous(
    breaks = c(0, 66.3, 120, 174, 221, 277, 33.15, 98.3, 145, 205, 249), 
    labels = c(0, "66  0", 54, "108  0", 47, 103,
               '\nX', '\n2L', '\n2R', '\n3L', '\n3R')) +
  theme(axis.ticks.x = element_blank()) +
  ylab(expression("-log"[10]*"(P-value)")) +
  xlab("Position (cM)")
```

## Cautionary Tale: Don't assume randomization will always work

  - Pooled sequencing of young and old flies: which haplotypes are associated with living a long time?
  - Roughly knew the population of possible genomes
  - Had the illumina reads from young pool and old pool

## Cautionary Tale: Don't assume randomization will always work

Sources of Sampling Error

  - Choosing 100 females for each pool
  - Prepping library & sequencing

## Approach 1: Use known population to estimate expectations from randomly choosing 100 females for each pool {.smaller}

<center>
<img src="https://i.imgur.com/F0UXoPC.png" width="90%" />
</center>

## Approach 2: Randomize the illumina read labels (young & old) and recalculate differences in haplotype frequencies {.smaller}

<center>
<img src="https://i.imgur.com/sO3AV1M.png" width="90%" />
</center>

## Turning flight in hummingbirds

<center>
<img src="https://i.imgur.com/Sacnb48.png" width="100%" />
</center>

## Turning flight in hummingbirds

<center>
<img src="https://i.imgur.com/Z144W14.png" width="80%" />
</center>

## Turning flight in hummingbirds

<center>
<img src="https://i.imgur.com/zo9Wuu0.png" width="80%" />
</center>

## Turning flight in hummingbirds

<center>
<img src="https://i.imgur.com/34njXoZ.png" width="80%" />
</center>

## No Quiz

End of Lectures
