---
title: "Non-independence: Introduction"
subtitle: "Quantitative Methods in Life Sciences"
author: 'Elizabeth King and Kevin Middleton'
output:
  ioslides_presentation:
    fig_width: 8
    css: styles.css
csl: evolution.csl
bibliography: Multivariate.bib
---

```{r setup, echo=FALSE, message=FALSE, warning=FALSE}
library(tidyverse)
library(cowplot)
library(readxl)
```


## IID is often violated

- Events are independent if the occurence of one does not influence the occurence of the other
- Assumption of most statistical tests
- In biological situations, independence assumption is often violated

<center>
<img src="https://i.imgur.com/2NrNYHZ.png" width="60%" />
</center>

## When is independence assumption violated?
  
- When there is a hierarchical structure to the data
    - e.g., several samples at each site 
- Shared ancestry
    - Phylogenetic relatedness (between populations/species)
    - Genetic relatedness (between individuals within populations/species)
- Spatial relationships    
- Sampling over time
- ...
  
## Why is non-independence a problem?

Does higher altitude drive the evolution of higher levels of hemoglobin?

```{r echo=FALSE, fig.width=7, fig.height=4}
ex.dat <- data.frame(Altitude = c(0, 1400, 2500, 3500, 4500, 6000),
                     Hemoglobin = c(9, 15, 11, 21, 16, 20),
                     Clade = rep(c('Low', 'High'),each = 3)) 
ggplot(ex.dat, aes(x = Altitude, y = Hemoglobin)) +
  geom_point(size = 6)
```

## Why is non-independence a problem?

Does higher altitude drive the evolution of higher levels of hemoglobin?

```{r echo=FALSE, fig.width=7, fig.height=4}
ggplot(ex.dat, aes(x = Altitude, y = Hemoglobin, color = Clade)) +
  geom_point(size = 6) +
  scale_color_manual(values = c("blue", "red")) +
  theme(legend.position = c(0.92, 0.17))
```

## Why is non-independence a problem?

**Pseudoreplication**: artificially inflated sample size lead to unfounded confidence

<center>
<img src="https://i.imgur.com/2AzYFXF.png" width="90%" />
</center>

## Conventional statistics assume

<center>
<img src="https://i.imgur.com/FeMvrT7.png" width="90%" />
</center>

## Best case for this example

<center>
<img src="https://i.imgur.com/kAf09JX.png" width="90%" />
</center>

## Genotype-Phenotype Relationships

Is a given genotype associated with a phenotype of interest?

- Consider two populations with ancestry differences
- Not accounting for ancestry differences can lead to spurious associations
- Population structure & relatedness are issues for all genetic association studies
    
## Genotype-Phenotype Relationships

Is a given genotype associated with a phenotype of interest?

```{r echo=FALSE, fig.width=6, fig.height=4}
set.seed(23141)

pop <- data.frame('population' = rep(c('A', 'B'), each = 20), 
                  'genotype' = c(c(rep(0, 14), rep(1, 6)),
                                 c(rep(0, 3), rep(1, 17))),
                  'phenotype' = c(rnorm(20, 5), rnorm(20, 3)))

ggplot(pop, aes(x = as.factor(genotype), y = phenotype)) +
  geom_point(position = position_jitter(0.05), alpha = 0.8) +
  labs(x = 'Genotype', y = "Phenotype") +
  stat_summary(fun.y = mean, geom = "point", size = 5,
               color = "grey60") + 
  stat_summary(fun.data = mean_se, geom = "errorbar",
               width = 0.05,
               color = "grey60", size = 1)
```

## Genotype-Phenotype Relationships

An effect of both 1) different allele frequency in two pops, and 2) different average trait value in two pops

```{r echo=FALSE, fig.width=7, fig.height=4}

ggplot(pop, aes(x = as.factor(genotype), y = phenotype,
                color = population)) +
  geom_point(position = position_jitter(0.05), alpha = 0.8) +
  labs(x = 'Genotype', y = "Phenotype") +
  stat_summary(fun.y = mean, geom = "point", size = 5,
               color = "grey60") + 
  stat_summary(fun.data = mean_se, geom = "errorbar",
               width = 0.05,
               color = "grey60", size = 1) +
  scale_color_manual(values = c("orange", "purple"))
```

## No Quiz

Lecture 11-2
