---
title: "Non-independence: Introduction"
subtitle: "Quantitative Methods in Life Sciences"
author: 'Elizabeth King and Kevin Middleton'
date: 'Last updated: `r Sys.Date()`'
output:
  ioslides_presentation:
    fig_width: 8
    css: styles.css
csl: evolution.csl
bibliography: Multivariate.bib
---

```{r setup, echo=FALSE, message=FALSE, warning=FALSE}
library(tidyverse)
library(cowplot)
library(readxl)
```

## Readings

- Garland T Jr, Adolph SC [-@Garland1994-vk] Why not to do two-species comparative studies: limitations on inferring adaptation. *Physiol Zool* 67:797-828.

## Resources

- Nunn CL. [-@Nunn2011-zi] *The Comparative Approach in Evolutionary Anthropology and Biology*. Chicago: University of Chicago Press.
- Nunn CL, Barton RA [-@Nunn2001-oq] Comparative methods for studying primate adaptation and allometry. *Evol Anthropol* 10:81â€“98.
- Yang J, Zaitlen NA, Goddard ME, Visscher PM, Price AL [-@Yang2014-cz] Advantages and pitfalls in the application of mixed-model association methods. *Nat Genet* 46:100-106.

## Even more resources

- http://www.anthrotree.info/: Website accompanying *The Comparative Approach in Evolutionary Anthropology and Biology*
- http://treethinkers.org/: UC Davis Bodega Bay Marine Lab Workshop in Applied Phylogenetics
- http://www.phylometh.org/: Brian O'Meara's (UT Knoxville) course on phylogenetic comparative methods

## IID is often violated

Recall IID = Independent and Identically Distributed
   
- Assumption of most statistical tests
- In biological situations, independence assumption is often violated

<center>
<img src="https://www.researchgate.net/profile/David_Reznick/publication/223968369/figure/fig2/AS:289284385984548@1445982180555/Figure-1-Map-of-the-Northern-Range-Mountains-of-Trinidad-showing-the-north-and-south.png" width="60%" />
</center>

## When is independence assumption violated?
  
- When there is a hierarchical structure to the data
    - e.g., several samples at each site 
- Shared ancestry
    - Phylogenetic relatedness (between populations/species)
    - Genetic relatedness (between individuals within populations/species)
- Spatial relationships    
- Sampling over time
- ...
  
## Why is non-independence a problem?

Does higher altitude drive the evolution of higher levels of hemoglobin?

```{r echo=FALSE, fig.width=7, fig.height=4}
ex.dat <- data.frame(Altitude = c(0, 1400, 2500, 3500, 4500, 6000),
                     Hemoglobin = c(9, 15, 11, 21, 16, 20),
                     Clade = rep(c('Low', 'High'),each = 3)) 
ggplot(ex.dat, aes(x = Altitude, y = Hemoglobin)) +
  geom_point(size = 6)
```

## Why is non-independence a problem?

Does higher altitude drive the evolution of higher levels of hemoglobin?

```{r echo=FALSE, fig.width=7, fig.height=4}
ggplot(ex.dat, aes(x = Altitude, y = Hemoglobin, color = Clade)) +
  geom_point(size = 6) +
  scale_color_manual(values = c("blue", "red")) +
  theme(legend.position = c(0.92, 0.17))
```

## Why is non-independence a problem?

**Pseudoreplication**: artificially inflated sample size lead to unfounded confidence

<center>
<img src="images/PCM_2.png" width="90%" />
</center>

## Conventional statistics assume

<center>
<img src="images/PCM_3.png" width="90%" />
</center>

## Best case for this example

<center>
<img src="images/PCM_4.png" width="90%" />
</center>

## Genotype-Phenotype Relationships

Is a given genotype associated with a phenotype of interest?

  - Consider two populations with ancestry differences
  - Not accounting for ancestry differences can lead to spurious associations
  - Population structure & relatedness are issues for all genetic association studies
    
## Genotype-Phenotype Relationships

Is a given genotype associated with a phenotype of interest?


```{r echo=FALSE, fig.width=6, fig.height=4}
set.seed(23141)

pop <- data.frame('population' = rep(c('A','B'),each=20), 
                  'genotype' = c(c(rep(0,14),rep(1,6)), c(rep(0,3),rep(1,17))),
                  'phenotype' = c(rnorm(20,5),rnorm(20,3)))

ggplot(pop, aes(x=as.factor(genotype),y=phenotype)) +
  geom_point(position=position_jitter(0.05), alpha=0.8) +
  xlab('genotype') +
  stat_summary(fun.y = mean, geom = "point", size = 5,
               color = "grey60") + 
  stat_summary(fun.data = mean_se, geom = "errorbar",
               width = 0.05,
               color = "grey60", size = 1)
  
```

## Genotype-Phenotype Relationships

An effect of both 1) different allele frequency in two pops, and 2) different average trait value in two pops

```{r echo=FALSE, fig.width=7, fig.height=4}

ggplot(pop, aes(x=as.factor(genotype),y=phenotype,color=population)) +
  geom_point(position=position_jitter(0.05),alpha=0.8) +
  xlab('genotype') +
  stat_summary(fun.y = mean, geom = "point", size = 5,
               color = "grey60") + 
  stat_summary(fun.data = mean_se, geom = "errorbar",
               width = 0.05,
               color = "grey60", size = 1)

  

```


    
## Quiz 11-1

Lecture 11-2

## References
