---
title: "Non-independence: Introduction"
subtitle: "Quantitative Methods in Life Sciences"
author: 'Elizabeth King and Kevin Middleton'
output:
  ioslides_presentation:
    fig_width: 8
    css: styles.css
csl: evolution.csl
bibliography: Multivariate.bib
---

```{r setup, echo=FALSE, message=FALSE, warning=FALSE}
library(knitr)
knitr::opts_chunk$set(cache = TRUE)
library(tidyverse)
library(cowplot)
library(readxl)

library(wesanderson)
library(ape)
library(phytools)
library(ggtree)
library(foreach)
library(reshape2)
library(geiger)
library(latex2exp)
library(nlme)
library(AICcmodavg)

theme_set(theme_cowplot())
```


## IID is often violated

- Observations are independent if the value or occurrence of one *does not* influence any others
- Standard assumption of most statistical tests
- In biological situations, independence assumption is often violated

<center>
<img src="https://i.imgur.com/2NrNYHZ.png" width="60%" />
</center>


## When is independence assumption violated?
  
- When there is a hierarchical structure to the data
    - e.g., several samples at each site 
- Shared ancestry
    - Phylogenetic relatedness (between populations/species)
    - Genetic relatedness (between individuals within populations/species)
- Spatial relationships    
- Repeat sampling over time
- ...
  
  
## Why is non-independence a problem?

Does higher altitude drive the evolution of higher levels of hemoglobin?

```{r echo=FALSE, fig.width=7, fig.height=4}
ex.dat <- data.frame(Altitude = c(0, 1400, 2500, 3500, 4500, 6000),
                     Hemoglobin = c(9, 15, 11, 21, 16, 20),
                     Clade = rep(c('Low', 'High'),each = 3)) 
ggplot(ex.dat, aes(x = Altitude, y = Hemoglobin)) +
  geom_point(size = 6)
```


## Why is non-independence a problem?

Does higher altitude drive the evolution of higher levels of hemoglobin?

```{r echo=FALSE, fig.width=7, fig.height=4}
ggplot(ex.dat, aes(x = Altitude, y = Hemoglobin, color = Clade)) +
  geom_point(size = 6) +
  scale_color_manual(values = c("blue", "red")) +
  theme(legend.position = c(0.82, 0.17))
```


## Why is non-independence a problem?

**Pseudoreplication**: artificially inflated sample size lead to unfounded confidence

<center>
<img src="https://i.imgur.com/2AzYFXF.png" width="90%" />
</center>


## Conventional statistics assume

<center>
<img src="https://i.imgur.com/FeMvrT7.png" width="90%" />
</center>


## Best case for this example

<center>
<img src="https://i.imgur.com/kAf09JX.png" width="90%" />
</center>


## Genotype-Phenotype Relationships

Is a given genotype associated with a phenotype of interest?

- Consider two populations with ancestry differences
- Not accounting for ancestry differences can lead to spurious associations
- Population structure & relatedness are issues for all genetic association studies
    
## Genotype-Phenotype Relationships

Is a given genotype associated with a phenotype of interest?

```{r echo=FALSE, fig.width=6, fig.height=4}
set.seed(23141)

pop <- data.frame('population' = rep(c('A', 'B'), each = 20), 
                  'genotype' = c(c(rep(0, 14), rep(1, 6)),
                                 c(rep(0, 3), rep(1, 17))),
                  'phenotype' = c(rnorm(20, 5), rnorm(20, 3)))

ggplot(pop, aes(x = as.factor(genotype), y = phenotype)) +
  geom_point(position = position_jitter(0.05), alpha = 0.8) +
  labs(x = 'Genotype', y = "Phenotype") +
  stat_summary(fun = mean, geom = "point", size = 5,
               color = "grey60") + 
  stat_summary(fun.data = mean_se, geom = "errorbar",
               width = 0.05,
               color = "grey60", size = 1)
```


## Genotype-Phenotype Relationships

An effect of both 1) different allele frequency in two pops, and 2) different average trait value in two pops

```{r echo=FALSE, fig.width=7, fig.height=4}
ggplot(pop, aes(x = as.factor(genotype), y = phenotype,
                color = population)) +
  geom_point(position = position_jitter(0.05), alpha = 0.8) +
  labs(x = 'Genotype', y = "Phenotype") +
  stat_summary(fun = mean, geom = "point", size = 5,
               color = "grey60") + 
  stat_summary(fun.data = mean_se, geom = "errorbar",
               width = 0.05,
               color = "grey60", size = 1) +
  scale_color_manual(values = c("orange", "purple"))
```


## Trees and tree thinking

- As soon as you have comparative data, you must consider the relationships among your taxa
    - No longer an option (30+ years of methods)
- Not concerned in this class with making the tree(s)

You have a tree and data
  - What do I do now?

## Where does your tree come from?

- Published trees
    - As is
    - Pruned down from published trees
- Constructed from multiple trees
- Including tree uncertainty
    - Compare effect of different taxon placements

<center>
*Trees are hypotheses of relationships. You have hypotheses resting on hypotheses.*
</center>

## Is body size associated with home range area in mammals?

Data from Garland et al. [-@Garland1992-bf]

- Body mass (km)
- Home range area (km^2^)
- "49lbr" data

## Home range vs. body mass

```{r echo=FALSE}
tree <- read.nexus("../data/49LBR70.PDI.nex")
lbr <- read_excel("../data/49LBR.xlsx")
lbr$Clade <- relevel(as.factor(lbr$Clade), ref = "Ungulates")

ggplot(lbr, aes(log_Mass, log_Home_Range)) +
  geom_point(size = 3) +
  labs(x = "log Body Mass", y = "log Home Range")
```

## Home range vs. body mass

```{r echo=FALSE}
ggplot(lbr, aes(log_Mass, log_Home_Range, color = Clade)) +
  geom_point(size = 3) +
  labs(x = "log Body Mass", y = "log Home Range") +
  theme(legend.position = c(0.85, 0.2)) +
  scale_color_manual(values = wes_palette("Moonrise2"))
```

## Home range vs. body mass

```{r echo=FALSE}
ggtree(tree, ladderize = FALSE) +
  geom_tiplab(size = 3) +
  geom_treescale()
```

## Home range vs. body mass

```{r echo=FALSE}
tree2 <- groupClade(tree, .node = 51)
ggtree(tree2, aes(color = group), ladderize = FALSE) +
  geom_tiplab(size = 3) +
  geom_treescale() +
  scale_color_manual(values = wes_palette("Moonrise2"))
```

## Why phylogenetic comparative methods?

1. Meet the assumptions of standard statistical tests
    - Any test you can do using "standard" methods has a phylogenetically informed counterpart.
    - What is your evolutionary model?
2. Ask and answer nuanced evolutionary questions
    - Trait evolution (continuous or categorical)
    - Trait co-evolution
    - Rates of evolution (identify nodes when rates shift significantly)
    - What is your evolutionary model?

## Branch lengths $\approx$ evolutionary model

"Real" branch lengths:

- Divergence dates (e.g., fossil calibrated)
- Genetic or character distance (relative)

Arbitrary branch lengths:

- All = 1
- Pagel's ultrametric transformation (1992): total height = deepest nesting

Either can be transformed.

## Divergence dated

```{r echo=FALSE}
ggtree(tree, ladderize = FALSE) +
  geom_tiplab(size = 3) +
  geom_treescale()
```

## All = 1

```{r echo=FALSE}
tree_1 <- tree
tree_1$edge.length <- rep(1, times = length(tree_1$edge.length))
ggtree(tree_1, ladderize = FALSE) +
  geom_tiplab(size = 3) +
  geom_treescale()
```

## Square-root divergence time

```{r echo=FALSE}
tree_sqrt <- tree
tree_sqrt$edge.length <- sqrt(tree_sqrt$edge.length)
ggtree(tree_sqrt, ladderize = FALSE) +
  geom_tiplab(size = 3) +
  geom_treescale()
```

## Models of evolution

- Brownian motion
- Ornstein-Uhlenbeck (OU)
- Variable rates (ACDC)
- Pagel's $\lambda$

**These models are all accomplished via branch length transformations.**

## Brownian motion

- The default, null underlying model for evolution
- Simplest, most parsimonious model of drift
- No selection
- At each step, a trait value is randomly altered based on a normal distribution (mean usually 0) with some variance.

## Brownian motion

Simulate 100 Brownian motion random walks with a length of 100 steps.

```{r}
set.seed(3)
nsim <- 100  # Number of simulations
t <- 0:100   # Time
s2 <- 0.01   # Variance

X <- foreach(i = 1:nsim, .combine = rbind) %do%
  c(0, cumsum(rnorm(n = length(t) - 1, sd = sqrt(s2))))
```

## Brownian motion

```{r echo=FALSE}
lims <- range(X)
M_one <- data.frame(t = 1:ncol(X),
                    x = as.numeric(X[1, ]))
ggplot(M_one, aes(t, x)) + geom_line() +
  ylim(lims) +
  labs(x = "Time (generations)", y = "Phenotype")
```

## Brownian motion

```{r echo=FALSE}
M <- as.data.frame(t(X))
M$t <- 1:ncol(X)
M_long <- melt(M, id = "t")
ggplot() +
  geom_line(data = M_long, aes(t, value, group = variable),
            alpha = 0.25) +
  geom_line(data = M_one, aes(t, x), color = "red", size = 2) +
  ylim(lims) +
  labs(x = "Time (generations)", y = "Phenotype")
```

## Brownian motion

```{r long_BM_sim, echo=FALSE, cache=TRUE}
set.seed(3)
nsim <- 500
t <- 0:1000
s2 <- 0.01

X <- foreach(i = 1:nsim, .combine = rbind) %do%
  c(0, cumsum(rnorm(n = length(t) - 1, sd = sqrt(s2))))

lims <- range(X)
M_one <- data.frame(t = 1:ncol(X),
                    x = as.numeric(X[1, ]))

M <- as.data.frame(t(X))
M$t <- 1:ncol(X)
M_long <- melt(M, id = "t")
ggplot() +
  geom_line(data = M_long, aes(t, value, group = variable),
            alpha = 0.25) +
  geom_line(data = M_one, aes(t, x), color = "red", size = 2) +
  ylim(lims) +
  labs(x = "Time (generations)", y = "Phenotype")
```

## Brownian motion

```{r echo=FALSE}
x <- as.numeric(t(M[nrow(M), 1:(ncol(M) - 1)]))
ggplot(data.frame(x), aes(x)) +
  geom_histogram(bins = 30) +
  labs(x = "Phenotype", y = "Count")
```

## Ornstein-Uhlenbeck model

- Based on Ornstein & Uhlenbeck [-@Uhlenbeck1930-xx]
- Brownian motion doesn't allow for adaptive evolution (drift only)
- Simplest model for an evolutionary process with selection (deterministic + stochastic)

> The OU model describes the motion of a species in the phenotypic space whereby the species moves randomly within the space, but is influenced by a central tendency such that large deviations from the central optimum receive a stronger force back toward the optimum [@Blomberg2003-jk]

## Ornstein-Uhlenbeck model (OU)

$$dX(t) = \alpha[\theta - X(t)]dt + \sigma dB(t)$$

- Change in trait X at time *t* is a function of selection $\alpha[\theta - X(t)]dt$ and drift $\sigma dB(t)$
    - $\alpha$ = strength of selection
    - $\theta$ = optimum
    - $\sigma$ = magnitude of drift
- $\alpha$ and one or more local optima ($\theta_1$, $\theta_2$, etc.) can be estimates
- Models compared with likelihood ratio tests, AIC, etc.

## Ornstein-Uhlenbeck model (OU)

Key readings:

- Martins E, Hansen TF [-@Hansen1996-ad] Phylogenies and the comparative method: A general approach to incorporating phylogenetic information into the analysis of interspecific data. *Am Nat* 149:646-667.
- Blomberg SP, Garland T Jr, Ives AR [-@Blomberg2003-jk] Testing for phylogenetic signal in comparative data: Behavioral traits are more labile. *Evolution* 57:717-745.
- Butler M, King A [-@Butler2004-ey] Phylogenetic comparative analysis: a modeling approach for adaptive evolution. *Am Nat* 164:683-695.

## OU evolution along a tree

```{r}
library(phytools)
set.seed(331)
tree <- pbtree(n = 15, scale = 10)
```

## OU evolution along a tree

```{r echo=FALSE}
ggtree(tree, ladderize = FALSE) +
  geom_tiplab(size = 3)
```

## OU evolution along a tree

$$dX(t) = \alpha[\theta - X(t)]dt + \sigma dB(t)$$

```{r}
x <- fastBM(tree,
            a = 0,           # Ancestral state at the root node
            theta = 3,       # Optimum
            alpha = 0.2,     # Strength of selection
            s2 = 0.1,      # Variance of the BM process
            internal = TRUE) # Return states for internal nodes
```

## OU evolution along a tree

```{r echo=FALSE}
phenogram(tree, x, spread.labels = TRUE, spread.cost = c(1, 0.01),
          main = "scale = 10, alpha = 0.2, s2 = 0.1")
```

## OU evolution along a tree

```{r echo=FALSE}
set.seed(331)
tree <- pbtree(n = 15, scale = 100)
x <- fastBM(tree,
            a = 0,           # Ancestral state at the root node
            theta = 3,       # Character state optimum
            alpha = 0.2,     # Strength of selection
            s2 = 0.1,      # Variance of the BM process
            internal = TRUE) # Return states for internal nodes
phenogram(tree, x, spread.labels = TRUE, spread.cost = c(1, 0.01),
          main = "scale = 100, alpha = 0.2, s2 = 0.1")
```

## OU evolution along a tree

```{r echo=FALSE}
set.seed(331)
tree <- pbtree(n = 15, scale = 10)
x <- fastBM(tree,
            a = 0,           # Ancestral state at the root node
            theta = 3,       # Character state optimum
            alpha = 0.5,     # Strength of selection
            s2 = 0.1,      # Variance of the BM process
            internal = TRUE) # Return states for internal nodes
phenogram(tree, x, spread.labels = TRUE, spread.cost = c(1, 0.01),
          main = "scale = 10, alpha = 0.5, s2 = 0.1")
```

## OU evolution along a tree

```{r echo=FALSE}
set.seed(331)
tree <- pbtree(n = 15, scale = 10)
x <- fastBM(tree,
            a = 0,           # Ancestral state at the root node
            theta = 3,       # Character state optimum
            alpha = 0.2,     # Strength of selection
            s2 = 0.5,      # Variance of the BM process
            internal = TRUE) # Return states for internal nodes
phenogram(tree, x, spread.labels = TRUE, spread.cost = c(1, 0.01),
          main = "scale = 10, alpha = 0.2, s2 = 0.5")
```

## Estimating an OU model

Rather than simulating data, estimate the values for $\alpha$ and $\theta$.

1. Given a tree and data (body mass in Caribbean anoles)
2. What model of evolution best fits the data?
    - Brownian motion
    - OU model with a single, clade-wide optimum
    - OU model with three body size optima (S/M/L)
    - OU model with four optima (including an ancestral size)
    - OU model based on multiple colonization events

## Estimating an OU model

Evolution toward two local optima:

<center>
<img src="https://i.imgur.com/8Iwp6A4.png" width="50%" />
</center>

## Estimating an OU model

<center>
<img src="https://i.imgur.com/qfuAvQ5.png" width="100%" />
</center>

## Estimating an OU model

<center>
<img src="https://i.imgur.com/vUwTJ2Z.png" width="100%" />
</center>

## Estimating an OU model

<center>
<img src="https://i.imgur.com/OFcpiST.png" width="75%" />
</center>

$\theta_{small} = `r round(exp(-1.4), 2)`$ mm head length.

## ACDC

- Evolution that either increases or decreases in rate over time.
    - Adaptive radiations followed by phenotypic stasis
    - Proposed by Blomberg et al. [-@Blomberg2003-jk]
- "Early burst" model of Harmon et al. [-@Harmon2010-gl]
- Didn't really ever catch on as much as OU.

## Models of evolution $\approx$ branch length transformations

Pagel's $\lambda$: how well does the tree fit the predicted covariance among the trait values

- $1$ = Brownian motion
- $>1$ = nodes pulled toward the tips (stronger covariance than expected by BM)
- $<1$ = nodes pulled toward the root (weaker covariance than expected by BM; more common)

$$\lim_{d\to0} = \mbox{"Star phylogeny"}$$

## Pagel's $\lambda$ branch scaling

```{r echo=FALSE}
set.seed(10)
tree <- pbtree(n = 30, b = 2)
ggtree(tree, ladderize = FALSE) +
  geom_tiplab(size = 3) +
  ggtitle(TeX("Original Branch Lengths ($\\lambda = 1$)"))
```

## Pagel's $\lambda$ branch scaling

```{r echo=FALSE}
ggtree(rescale(tree, "lambda", 0.5), ladderize = FALSE) +
  geom_tiplab(size = 3) +
  ggtitle(TeX("$\\lambda = 0.5$"))
```

## Pagel's $\lambda$ branch scaling

```{r echo=FALSE}
ggtree(rescale(tree, "lambda", 0.1), ladderize = FALSE) +
  geom_tiplab(size = 3) +
  ggtitle(TeX("$\\lambda = 0.1$"))
```

## Pagel's $\lambda$ branch scaling

```{r echo=FALSE}
ggtree(rescale(tree, "lambda", 0.001), ladderize = FALSE) +
  geom_tiplab(size = 3) +
  ggtitle(TeX("$\\lambda = 0.001$"))
```

## Home range vs. body mass

```{r echo=FALSE}
tree <- read.nexus("../data/49LBR70.PDI.nex")
lbr <- read_excel("../data/49LBR.xlsx") %>% 
  as.data.frame()
lbr$Clade <- relevel(as.factor(lbr$Clade), ref = "Ungulates")
row.names(lbr) <- lbr$Species

# Clean up tree
tree$tip.label <- gsub("\\.", "", tree$tip.label)
tree$tip.label <- gsub("\\:", "", tree$tip.label)
```

```{r echo=FALSE}
tree2 <- groupClade(tree, .node = 51)
ggtree(tree2, aes(color = group), ladderize = FALSE) +
  geom_tiplab(size = 3) +
  geom_treescale() +
  scale_color_manual(values = wes_palette("Moonrise2"))
```

## Home range vs. body mass

```{r echo=FALSE}
ggplot(lbr, aes(log_Mass, log_Home_Range, color = Clade)) +
  geom_point(size = 3) +
  labs(x = "log Body Mass", y = "log Home Range") +
  theme(legend.position = c(0.85, 0.2)) +
  scale_color_manual(values = wes_palette("Moonrise2"))
```

## Working with `nlme::gls()`

- We used `gls()` to fit the squid model, allowing correlated variance on a per-month basis
- Phylogenetically based correlation variation is a similar issue
- Choose a phylogenetic correlation structure

## Phylogenetic correlation structures

See ?`corClasses`

- `corBrownian`: BM model of correlation
- `corPagel`: Pagel's $\lambda$
- `corMartins`: OU model originally proposed by Martins and Hansen [-@Martins1997-yv]
- `corBlomberg`: ACDC model of Blomberg et al. [-@Blomberg2003-jk]

## Model 1: Mass, BM

```{r eval=FALSE}
library(nlme)
fm1 <- gls(log_Home_Range ~ log_Mass, data = lbr,
           correlation = corBrownian(phy = tree))
summary(fm1)
```

## Model 1: Mass, BM {.smaller}

```{r echo=FALSE}
fm1 <- gls(log_Home_Range ~ log_Mass, data = lbr,
           correlation = corBrownian(phy = tree))
summary(fm1)
```

## Model 2: Mass + Clade, BM {.smaller}

```{r}
fm2 <- gls(log_Home_Range ~ log_Mass + Clade , data = lbr,
           correlation = corBrownian(phy = tree))
summary(fm2)
```

## Model 3: Mass * Clade, BM {.smaller}

```{r}
fm3 <- gls(log_Home_Range ~ log_Mass * Clade , data = lbr,
           correlation = corBrownian(phy = tree))
summary(fm3)
```

## Model 4: Mass, OU {.smaller}

Estimating OU parameter alpha often fails (`Error in corFactor.corStruct(object)`)

- Brute-force optimization of OU alpha parameter
- Maximize the log-likelihood for a range of models with different values of alpha.

```{r}
myOU <- function(alpha, dat, tree){
  fm <- gls(log_Home_Range ~ log_Mass, data = dat, 
            correlation = corMartins(alpha, tree, fixed = TRUE))
  return(as.numeric(fm$logLik))
}

alpha <- optimize(f = myOU, interval = c(0, 2), 
                  dat = lbr, tree = tree, maximum = TRUE)
alpha
```

## Model 4: Mass, OU

Fit a model with the alpha parameter fixed at the optimized value:

```{r}
ou_mod <- corMartins(value = alpha$maximum,
                     phy = tree, fixed = TRUE)
fm4 <- gls(log_Home_Range ~ log_Mass, data = lbr,
           correlation = ou_mod)
```

## Model 4: Mass, OU {.smaller}

```{r echo=FALSE}
summary(fm4)
```

## Model comparison with AICc

Models 5 and 6 ANCOVA with OU. Models 7-9, estimate Pagel's $\lambda$.

```{r echo=FALSE}
myOU <- function(alpha, dat, tree){
  fm <- gls(log_Home_Range ~ log_Mass + Clade, data = dat, 
            correlation = corMartins(alpha, tree, fixed = TRUE))
  return(as.numeric(fm$logLik))
}

alpha <- optimize(f = myOU, interval = c(0, 2), 
                  dat = lbr, tree = tree, maximum = TRUE)
ou_mod <- corMartins(value = alpha$maximum,
                     phy = tree, fixed = TRUE)
fm5 <- gls(log_Home_Range ~ log_Mass + Clade, data = lbr,
           correlation = ou_mod)

myOU <- function(alpha, dat, tree){
  fm <- gls(log_Home_Range ~ log_Mass * Clade, data = dat, 
            correlation = corMartins(alpha, tree, fixed = TRUE))
  return(as.numeric(fm$logLik))
}

alpha <- optimize(f = myOU, interval = c(0, 2), 
                  dat = lbr, tree = tree, maximum = TRUE)
ou_mod <- corMartins(value = alpha$maximum,
                     phy = tree, fixed = TRUE)
fm6 <- gls(log_Home_Range ~ log_Mass * Clade, data = lbr,
           correlation = ou_mod)

fm7 <- gls(log_Home_Range ~ log_Mass, data = lbr,
           correlation = corPagel(1, phy = tree))
fm8 <- gls(log_Home_Range ~ log_Mass + Clade, data = lbr,
           correlation = corPagel(1, phy = tree))
fm9 <- gls(log_Home_Range ~ log_Mass * Clade, data = lbr,
           correlation = corPagel(1, phy = tree))

mods <- matrix(NA, nrow = 3, ncol = 3)
row.names(mods) <- c("Brownian Motion", "Stabilizing Selection (OU)", "Pagel's $\\lambda$")
colnames(mods) <- c("Mass", "Mass + Clade", "Mass * Clade")
mods[1, 1] <- AICc(fm1)
mods[1, 2] <- AICc(fm2)
mods[1, 3] <- AICc(fm3)
mods[2, 1] <- AICc(fm4)
mods[2, 2] <- AICc(fm5)
mods[2, 3] <- AICc(fm6)
mods[3, 1] <- AICc(fm7)
mods[3, 2] <- AICc(fm8)
mods[3, 3] <- AICc(fm9)
kable(round(mods, 1))
```

## Best performing model {.smaller}

OU model:

$$Home~Range \sim Mass + Clade$$

```{r echo=FALSE}
summary(fm5)
```


## Kinship (non-independence) as a tool

When modeled correctly, non-independence between samples can be very useful:

- Forms the basis for how we estimate heritabilities, genetic correlations, etc.
- Very similar in concept to repeatability
- Estimate the variance within families/lines (groups) versus between families/lines
- Modern methods can consider complex pedigrees or relatedness estimated directly from genomic data

## Heritability and genetic correlation

<center>
<img src="https://i.imgur.com/AquXKrw.jpg" width="65%" />
</center>

## Load pedigree

```{r, message=FALSE}
library(MCMCglmm)
library(kinship2)

data(BTped)
str(BTped)
```

## Look at part of pedigree

```{r echo=FALSE}
Nped <- BTped[which(apply(BTped, 1, function(x) {
  any(x == "R187920" | x == "R187921")})), ]
Nped
```

## Plot part of pedigree

```{r warning=FALSE, echo=FALSE}
ped <- pedigree(id = Nped$animal,
                dadid = Nped$sire,
                momid = Nped$dam,
                sex = c(2, 1, rep(3, nrow(Nped) - 2)))
plot(ped)
```

## Genetic distance

```{r echo=FALSE}
Aped <- 2 * kinship2::kinship(Nped[, 1], Nped[, 2], Nped[, 3])
Aped[1:6,1:6]
```

## Data

```{r}
data(BTdata)
head(BTdata)
```

## Heritability of tarsus length

```{r, eval = FALSE}
prior <- list(R = list(V = 0.002, n = 2),
              G = list(G1 = list(V = 0.002, n = 2)))

fm1 <- MCMCglmm(
  fixed = tarsus ~ 1,
  random = ~ animal,
  prior = prior,
  family = c("gaussian"),
  nitt = 60000, burnin = 10000, thin = 25,
  data = BTdata, pedigree = BTped,
  verbose = FALSE)
save(fm1, file = "../data/BT_herit.Rda")
```

```{r echo=FALSE}
load("../data/BT_herit.Rda")
```

## Fixed effects

```{r}
plot(fm1$Sol)
```

## Random effects

```{r}
plot(fm1$VCV)
```

## Heritability

$$h^2 = \frac{V_a}{V_a + V_r}$$

```{r}
h2 <- fm1$VCV[, "animal"] / 
  (fm1$VCV[, "animal"] + fm1$VCV[, "units"])
```

```{r echo=FALSE, fig.height=3}
plot(h2)
```

## Heritability

```{r}
median(h2)
HPDinterval(h2)
```

## Genetic correlation

```{r eval=FALSE}
prior <- list(R = list(V = diag(2) * 1.001, n = 2),
              G = list(G1 = list(V = diag(2)  * 1.001, n = 2)))

fm2 <- MCMCglmm(
  fixed = cbind(back, tarsus) ~ trait,
  random = ~ us(trait):animal,
  rcov = ~ us(trait):units,
  prior = prior,
  family = c("gaussian", "gaussian"),
  nitt = 60000, burnin = 10000, thin = 25,
  data = BTdata, pedigree = BTped,
  verbose = FALSE)
save(fm2, file = "../data/BT_genetic_correlation.Rda")
```

```{r echo=FALSE}
load("../data/BT_genetic_correlation.Rda")
```

## Genetic correlation

```{r}
Genet_Corr <- fm2$VCV[ , "traittarsus:traitback.animal"] /
  sqrt(fm2$VCV[ , "traittarsus:traittarsus.animal"] *
         fm2$VCV[ , "traitback:traitback.animal"])
```

```{r echo=FALSE}
plot(Genet_Corr)
```

## Genetic correlation

```{r}
median(Genet_Corr)
HPDinterval(Genet_Corr)
```

## Kinship and GWAS

When individuals are related in different ways, correcting for kinship is critical to avoid spurious associations. This is necessary even when not including close relatives (e.g., siblings, cousins, etc.)

Typical data:

  - Hundreds of thousands to millions of SNPs
  - Phenotypes
  - Hundreds to thousands of individuals
  - That's **a lot** of tests! (speed matters)

## Kinship and GWAS

The typical model is a Linear Mixed Model (LMM) with fixed and random effects.

$$Y = W\beta + G\gamma + g + e$$
  
*Y* is the phenotype

Fixed effects:

  - *W* is a covariate (or more than one) with $\beta$ as the coefficient(s)
  - *G* is the genotype with $\gamma$ as the coefficient

## Kinship and GWAS

The typical model is a Linear Mixed Model (LMM) with fixed and random effects.

$$Y = W\beta + G\gamma + g + e$$
  
Random effects:

  - *g* is a vector of polygenic effects scaled according to relatedness between individuals
  - *e* is the error term

## Kinship Matrix

Kinship matrix can be 
  
  - Estimated from a pedigree
  - Estimated from the genomic data

## Mice on Meth

<center>
<img src="https://i.imgur.com/kaeCImN.jpg" width="85%" />
</center>

## Mice on Meth

<center>
<img src="https://i.imgur.com/Rst1DwD.jpg" width="90%" />
</center>
